Limit overall number of threads used by TBB

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22062 Fix sequential MKL case
* #22051 Update intra_inter_benchmark
* #22047 Experimental option to use single thread pool
* **#22045 Limit overall number of threads used by TBB**

Summary:
Using global control to set a limit on number of threads used by TBB

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_OPENMP=0 USE_TBB=1 MKL_SEQ=1
MKLDNN_THREADING=SEQ USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1
python setup.py develop --cmake

./build/bin/parallel_info
./build/bin/thread_init_test
./build/bin/test_parallel

Differential Revision: [D15930319](https://our.internmc.facebook.com/intern/diff/D15930319)

Commit:
Limit overall number of threads used by TBB

Summary:
Using global control to set a limit on number of threads used by TBB

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_OPENMP=0 USE_TBB=1 MKL_SEQ=1
MKLDNN_THREADING=SEQ USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1
python setup.py develop --cmake

./build/bin/parallel_info
./build/bin/thread_init_test
./build/bin/test_parallel

cpuhrsch commented on aten/src/ATen/ParallelNativeTBB.cpp:
```diff
@@ -6,6 +6,8 @@
 #include <mutex>
 
 #include "tbb/tbb.h"
+#define TBB_PREVIEW_GLOBAL_CONTROL 1
```
The only thing that makes me a bit queasy here is that this is an experimental feature.

ilia-cher replied:
well, yes, but this is also in official docs (https://software.intel.com/en-us/node/589744)
and that's the only setting I know that actually works

ilia-cher replied:
the example in https://software.intel.com/en-us/node/589744 actually doesn't have "#define TBB_PREVIEW_GLOBAL_CONTROL 1" so I guess it might not be so experimental in the latest version

ilia-cher replied:
also there's no mention of TBB_PREVIEW_GLOBAL_CONTROL in current intel/tbb repo

Commit:
Update on "Limit overall number of threads used by TBB"

Limit overall number of threads used by TBB

Summary:
Using global control to set a limit on number of threads used by TBB

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_OPENMP=0 USE_TBB=1 MKL_SEQ=1
MKLDNN_THREADING=SEQ USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1
python setup.py develop --cmake

./build/bin/parallel_info
./build/bin/thread_init_test
./build/bin/test_parallel

gh-metadata: pytorch pytorch 22045 gh/ilia-cher/29/head

cpuhrsch reviewed (approved):
Buyer beware

Commit:
Update on "Limit overall number of threads used by TBB"

Limit overall number of threads used by TBB

Summary:
Using global control to set a limit on number of threads used by TBB

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_OPENMP=0 USE_TBB=1 MKL_SEQ=1
MKLDNN_THREADING=SEQ USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1
python setup.py develop --cmake

./build/bin/parallel_info
./build/bin/thread_init_test
./build/bin/test_parallel

gh-metadata: pytorch pytorch 22045 gh/ilia-cher/29/head
